File: Yandex/Practicum/Algorightms_and_Data_Structures/1_sprint/K.Списочная_форма/solution.cpp
#include <iostream>
#include <vector>


void solution(int xLength, std::vector<int>& x, int k) {
    /*
    Time Complexity O(N), where N is the length of the largest number
    Memory Complexity O(N), where N is the length of the largest number
    */
    std::vector<int> output;
    int addition = 0;
    for (int index = xLength - 1; index >= 0; --index) {
        int sumDigits = x[index] + k % 10 + addition;
        addition = sumDigits / 10;
        k /= 10;
        output.push_back(sumDigits % 10);
    }
    while (k) {
        int sumDigits = k % 10 + addition;
        k /= 10;
        addition = sumDigits / 10;
        output.push_back(sumDigits % 10);
    }
    if (addition) {
        output.push_back(addition);
    }
    int n = output.size();
    for (int index = n - 1; index >= 0; --index) {
        std::cout << output[index];
        if (index) {
            std::cout << " ";
        }
    }
    std::cout << "\n";
}


int main() {
    int xLength;
    std::cin >> xLength;
    std::vector<int> x;
    for (int i = 0; i < xLength; ++i) {
        int digit;
        std::cin >> digit;
        x.push_back(digit);
    }
    int k;
    std::cin >> k;
    solution(xLength, x, k);
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/1_sprint/K.Списочная_форма/solution.py
import sys


def solution(x_length: int, x: list[int], k: int) -> None:
    """
    Time Complexity O(N), where N is the length of place values of the largest number
    Memory Complexity O(N), where N is the length of place values of the largest number
    """
    output: list[int] = []
    additional: int = 0
    for idx in range(x_length - 1, -1, -1):
        sum_digits: int = x[idx] + k % 10 + additional
        additional = sum_digits // 10
        output.append(sum_digits % 10)
        k //= 10
    while k:
        sum_digits: int = k % 10 + additional
        additional = sum_digits // 10
        output.append(sum_digits % 10)
        k //= 10
    if additional:
        output.append(additional)
    n: int = len(output)
    for idx in range(n - 1, -1, -1):
        sys.stdout.write(str(output[idx]))
        if idx:
            sys.stdout.write(' ')
    sys.stdout.write('\n')


if __name__ == '__main__':
    x_length: int = int(sys.stdin.readline().rstrip())
    x: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    k: int = int(sys.stdin.readline().rstrip())
    solution(x_length, x, k)
